### PR TITLE
EOS-11639 : Memory leak in cfs_fs_delete()

### PR DESCRIPTION
**Pull request by shahidSTX**
# cortx-fs Change Summary

## Problem Statement
_[EOS-11639](https://jts.seagate.com/browse/EOS-11639):_
_Memory leak in cfs_fs_delete_

## Problem Description
_cfs_fs_node and its contents are not freed, need to make free calls_

## Solution Overview
_Making appropriate free calls to free root_node, cfs_fs and cfs_fs_node_

## Unit Test Cases
_Able to create FS, mount FS, run basic I/O and delete FS_

<testing>

```
[root@ssc-vm-c-0688 cortx-posix]# mount -o vers=4.0 127.0.0.1:/fs1 /mnt/dir1
[root@ssc-vm-c-0688 cortx-posix]# echo "random text" > /mnt/dir1/testfile
[root@ssc-vm-c-0688 cortx-posix]# cat /mnt/dir1/testfile
random text
[root@ssc-vm-c-0688 cortx-posix]# umount /mnt/dir1

[root@ssc-vm-c-0688 cortx-posix]# cortxfscli fs create testFS
Created
[root@ssc-vm-c-0688 cortx-posix]# cortxfscli fs list
OK
FS ID   FS Name                             Exported        Protocol    Export Options
1       fs1                                 YES             nfs         {"proto": "nfs", "secType": "sys", "Filesystem_id": "192.1", "client": "1", "clients": "*", "Squash": "no_root_squash", "access_type": "RW", "protocols": "4", "pnfs_enabled": "false", "data_server": "10.230.241.28"}
2       testFS                              NO              None        ""
[root@ssc-vm-c-0688 cortx-posix]#
[root@ssc-vm-c-0688 cortx-posix]# cortxfscli fs delete testFS
OK

```

</testing>

## Testing
<details>
  <summary>Click here to see the test execution output</summary>

```
[root@ssc-vm-c-0688 cortx-posix]# ./scripts/test.sh
 --
 --
Configuration Completed
Clean indexes prepared
NSAL Unit tests
NS Tests
Test results are logged to /var/log/cortx/test/ut/ut_nsal.log
Total tests  = 5
Tests passed = 5
Tests failed = 0

Iterator test
Test results are logged to /var/log/cortx/test/ut/ut_nsal.log
Total tests  = 2
Tests passed = 2
Tests failed = 0

KVTree test
Test results are logged to /var/log/cortx/test/ut/ut_nsal.log
Total tests  = 17
Tests passed = 17
Tests failed = 0

CORTXFS Unit tests
Endpoint ops Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.logs
Total tests  = 4
Tests passed = 4
Tests failed = 0

FS Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 3
Tests passed = 3
Tests failed = 0

Directory tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 14
Tests passed = 14
Tests failed = 0

File creation tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 4
Tests passed = 4
Tests failed = 0

Link Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 6
Tests passed = 6
Tests failed = 0

Rename tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 3
Tests passed = 3
Tests failed = 0

Attribute Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 5
Tests passed = 5
Tests failed = 0

Xattr file Tests
Test results are logged to /var/log/cortx/test/ut/xattr_file_ops.log
Total tests  = 9
Tests passed = 9
Tests failed = 0

Xattr dir Tests
Test results are logged to /var/log/cortx/test/ut/xattr_dir_ops.log
Total tests  = 9
Tests passed = 9
Tests failed = 0

IO tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 6
Tests passed = 6
Tests failed = 0

DSAL Unit tests
Dsal basic test
Test results are logged to /var/log/cortx/test/ut/ut_dsal.log
Total tests  = 11
Tests passed = 11
Tests failed = 0

Dsal IO test
Test results are logged to /var/log/cortx/test/ut/ut_dsal.log
Total tests  = 1
Tests passed = 1
Tests failed = 0

Dsal space stats test
Test results are logged to /var/log/cortx/test/ut/ut_dsal.log
Total tests  = 2
Tests passed = 2
Tests failed = 0

[root@ssc-vm-c-0688 cortx-posix]# ./scripts/test.sh -t cortxfs
 --
 --
Configuration Completed
Clean indexes prepared
CORTXFS Unit tests
Endpoint ops Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.logs
Total tests  = 4
Tests passed = 4
Tests failed = 0

FS Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 3
Tests passed = 3
Tests failed = 0

Directory tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 14
Tests passed = 14
Tests failed = 0

File creation tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 4
Tests passed = 4
Tests failed = 0

Link Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 6
Tests passed = 6
Tests failed = 0

Rename tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 3
Tests passed = 3
Tests failed = 0

Attribute Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 5
Tests passed = 5
Tests failed = 0

Xattr file Tests
Test results are logged to /var/log/cortx/test/ut/xattr_file_ops.log
Total tests  = 9
Tests passed = 9
Tests failed = 0

Xattr dir Tests
Test results are logged to /var/log/cortx/test/ut/xattr_dir_ops.log
Total tests  = 9
Tests passed = 9
Tests failed = 0

IO tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 6
Tests passed = 6
Tests failed = 0
```

</details>

## Checklist
- [x] **Compilation:** _This patch does not break compilation_
- [x] **Merge conflicts:** _None_
- [ ] **Code review:** _In progress_
- [X] **Sanity Testing:** _Done as mentioned in testing above_
- [ ] **Documentation:** _TBD_

### CODE REVIEW

**Comment by s-arya** [comment]
Could you please add the details of the testing that was performed to verify this fix?
Also, please add the description as per the template in https://github.com/Seagate/cortx-fs/pull/19 viz.  Problem Statement, Problem Description, Fix, and Testing.  Also, please add the checklist.

**Comment by pratyush-seagate** [comment]
> 
> 
> Could you please add the details of the testing that was performed to verify this fix?
> Also, please add the description as per the template in #19 viz. Problem Statement, Problem Description, Fix, and Testing. Also, please add the checklist.



> 
> 
> Could you please add the details of the testing that was performed to verify this fix?
> Also, please add the description as per the template in #19 viz. Problem Statement, Problem Description, Fix, and Testing. Also, please add the checklist.

Yes I'm yet to publish this review formally, once all the formats are in place, will publish it in the team's review queue.

**Comment by pratyush-seagate** [comment]
Please excuse me for putting comment on wrong PR thinking the comment meant for one of my PR. Please ignore it.

**Review by sachinpunadikar — APPROVED** [review]
+1

**Review by gshipra — COMMENTED** [review]
We need to do more rearrangement of code in this function.
Need to deallocate all members of fs_node and do kvnode_fini.
Please look at this function: cfs_fs_fini()

**Comment by shahidSTX** [comment]
> We need to do more rearrangement of code in this function.
> Need to deallocate all members of fs_node and do kvnode_fini.
> Please look at this function: cfs_fs_fini()

Thanks Shipra for pointing out to cfs_fs_fini(). The changes that I have made ensure that all the members of fs_node are deallocated, some already done and rest are done here. After these changes, we are functionally doing the same things here for one file system exactly like what is being done by cfs_fs_fini for all file systems. But I agree that re-arranging the code would be more readable, I myself had to go through some difficulty understanding which members are already deallocated and which are yet to be.

**Comment by shahidSTX** [comment]
Addressed Shipra's comments and rearranged the code, UT complete and status updated. Please continue reviewing.

File: src/fs/fs.c [hunk 1]
```diff
@@ -545,19 +545,28 @@ int cfs_fs_delete(const str256_t *fs_name)
...
 	}
 
+	/* Remove fs from the cortxfs list */
+	fs_node = container_of(fs, struct cfs_fs_node, cfs_fs);
+	LIST_REMOVE(fs_node, link);
+
+	/* Call fini for all members of fs */
 	RC_WRAP_LABEL(rc, out, cfs_ino_num_gen_fini, fs);
 	RC_WRAP_LABEL(rc, out, kvtree_fini, fs->kvtree);
+	kvnode_fini(fs->root_node);
```
Comment (gshipra) [inline src/fs/fs.c:555]: You need to do following 👍 
tenant_free
kvnode_fini
kvtree_fini
 free(fs_node->cfs_fs.ns);
 free(fs_node->cfs_fs.root_node);
 free(fs_node->cfs_fs.kvtree);
free(fs_node)

Comment (gshipra) [reply src/fs/fs.c:555]: kvnode_fini happens before kvtree_fini.
Few lines of this are going to be part of single function which I am working on. For the sake of this defect you can do this in order. I will change it to one line call as part of my changes.

If you have doubts, you can check with @pratyush-seagate , he has worked on this code before.

**Comment by shahidSTX** [comment]
Addresses the review comments with the help of @gshipra , testing done same as above and verified.

**Review by gshipra — APPROVED** [review]

**Review by pratyush-seagate — APPROVED** [review]
After the discussion with @shahidSTX it seems to be that the delete path should be calling the finish code first for that specific file server. The code for finish operation should be modular and delete should do the final cleanup (including any on-disk remaining). I think this code can be refactored in future for better readability and modularity. But for the purpose of this ticket, the change looks good to me.

**Pull request merged**